File: algorithms/backtracking/sum-of-subsets/SumOfSubsets.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;

class SumOfSubsets {


    public static void main(String[] args){
        System.out.println("Sum of Subsets Algorithm using Java");

        ArrayList<Integer> weights = new ArrayList<>(Arrays.asList(90,38,76,58,87,62,63,98,26,41));
        int requiredSum = 200;
        int remainingSum = 0;
        for(int i=0; i<weights.size(); i++){
            remainingSum += weights.get(i);
        }
        int[] solutionVector = new int[weights.size()];

        Collections.sort(weights);

        sumOfSubsets(0,1,remainingSum,requiredSum,solutionVector,weights);

        return;
    }


    public static void sumOfSubsets(int s, int k, int r, int m, int[] x, ArrayList<Integer> w){
        x[k-1] = 1;

        if(s+w.get(k-1) == m){
            print_solution(x,w);
        }else if(s+w.get(k-1)+w.get(k) <= m){
            sumOfSubsets(s+w.get(k-1), k+1, r-w.get(k-1), m, x, w);
        }

        if(s+r-w.get(k-1) >= m && s+w.get(k) <= m){
            x[k-1] = 0;
            sumOfSubsets(s, k+1, r-w.get(k-1), m, x, w);
        }

        return;
    }

    public static void print_solution(int[] x, ArrayList<Integer> w){
        int n = x.length;

        System.out.print("[ ");
        for(int i=0; i<n-1; i++ ){
            System.out.print(x[i]*w.get(i)+", ");
        }
        System.out.println(x[n-1]*w.get(n-1)+" ]");

        return;
    }


}

File: algorithms/backtracking/sum-of-subsets/sum_of_subsets.py
def main() -> None:
    print("Sum of Subsets Algorithm using python") 

    weights: list[int] = [90,38,76,58,87,62,63,98,26,41]
    weights.sort()
    solution_vector: list[int] = [0 for _ in range(len(weights))]
    required_sum = 200
    remaining_sum = sum(weights)

    sum_of_subsets(0,1,remaining_sum,required_sum,solution_vector,weights)

    return


def sum_of_subsets(s: int, k: int, r: int, m: int, x: list[int], w: list[int]) -> None:
    
    x[k-1] = 1

    if s+w[k-1] == m:
        print([x[i]*w[i] for i in range(len(x))])
    elif s+w[k-1]+w[k] <= m:
        sum_of_subsets(s+w[k-1],k+1,r-w[k-1],m,x,w)

    if s+r-w[k-1] >= m and s+w[k] <= m:
        x[k-1] = 0
        sum_of_subsets(s,k+1,r-w[k-1],m,x,w)

    return 


if __name__ == "__main__":
    main()